docs/RFCS/20190318_error_handling.md

```diff
@@ -0,0 +1,2090 @@
+3. the resulting error is sent across the wire.
+
+When this occurs, the library makes extra effort to convert the
+wrapper object from `github.com/pkg/errors` into a form that's
+encodable, so as to preserve all the chain of causes and the
+intermediate message prefixes added via `github.com/pkg/errors.Wrap()`.
+
+### Barrier errors
+
+When an assertion failure is encountered while handling an error, we want
+to mark the assertion failure to be an "unrecoverable error" and let it flow
+to the top level (and telemetry) without any form of further handling.
+
+This type of error must inhibit any other mechanism that inspects causes.
+
+For this purpose the library provides a special error type:
+`BarrierError`.  This contains an optional "internal cause" however
+*this cause is not visible to the `Cause()` and `Unwrap()` methods*
+(nor the other mechanisms provided below). This original cause is only
+visible:
+
+- when inspecting the error object, e.g. via `%+v` formatting;
+- when the error flows out of the system (e.g. towards a SQL client);
+- when reported to telemetry.
+
+The following functions create barrier errors:
+
+- `errors.AssertionFailed()`
+- `errors.NewAssertionErrorWithWrappedErr()`
+
+Additionally, a barrier error interacts with the pg error code as follows:
+
+- if the original pg error code was absent or was for a "normal" error,
+  the barrier error will mask the original code and replace it
+  with `XX000/CodeInternalError`.
+
+- if the original pg error code was present with the `XX` prefix, then
+  the original code is preserved. This covers the following situation:
+
+  1. a low level storage failure is encountered and generates
+     `CodeDataCorruptedError` (`XX0001`)
+  2. the storage failure trips an unexpected path in the code (an
+     assertion failure.)
+
+  In that case we wish that the assertion failure preserves the
+  serious original code.
+
+### Promotion of `XX` codes to barrier errors
+
+When a `XX` code is provided via `Wrapx`, `WithDefaultCode`, etc, the
+result becomes a barrier and the original error, if any, is masked as an
+"internal cause" invisible via `causer`/`Wrapper` interfaces.
+
+This makes it possible to e.g. wrap with `CodeCCLRequired` (`XXC01`)
+on a CCL-only path and ensure that the original error is prevented
+from triggering further processing (e.g. retries).
+
+### Multiple causes
+
+CockroachDB contains multiple code patterns that try something, then
+if that first something results in an error try something else.
+
+If the second action itself results in error, there are then *two* error objects.
+
+Prior to this RFC, one of the errors would be "dropped on the floor"
+or, at best, flattened into a text message with
+e.g. `errors.Wrapf(err1, "while handling %v", err2)`.
+
+The proposed library extends this behavior and makes it possible to
+preserve multiple causes using `WithOtherCause()`, for example:
```

Comment:
This seems like it has to be potential to be overly complex. Is it really that important to preserve the other cause? What's wrong with the current behavior in practice?